src/coreclr/vm/frozenobjectheap.cpp

```diff
@@ -135,34 +150,48 @@ FrozenObjectSegment::FrozenObjectSegment(size_t sizeHint) :
         ThrowOutOfMemory();
     }
 
+    m_pStart = static_cast<uint8_t*>(committedAlloc);
+    m_pCurrent = m_pStart + sizeof(ObjHeader);
+    m_SizeCommitted = FOH_COMMIT_SIZE;
+    INDEBUG(m_ObjectsCount = 0);
+
     // ClrVirtualAlloc is expected to be PageSize-aligned so we can expect
     // DATA_ALIGNMENT alignment as well
     _ASSERT(IS_ALIGNED(committedAlloc, DATA_ALIGNMENT));
+}
 
-    segment_info si;
-    si.pvMem = committedAlloc;
-    si.ibFirstObject = sizeof(ObjHeader);
-    si.ibAllocated = si.ibFirstObject;
-    si.ibCommit = FOH_COMMIT_SIZE;
-    si.ibReserved = m_Size;
+void FrozenObjectSegment::RegisterOrUpdate()
+{
+    m_SizeCommittedRegistered = m_SizeCommitted;
```

Comment:
`m_SizeCommitted` and `m_pCurrent` can be updated by the other thread while they are being read here. Can it lead to race conditions?

Reply:
I think you may want to pass how far the current thread got as an argument to `RegisterOrUpdate`. `RegisterOrUpdate` can then decide whether some other thread beat it to register/update the space, or whether the space still needs to registering/updating.

Reply:
Thanks! Indeed could be a race condition, fixed